### PR TITLE
net: ipv4: Fix ARP probe check in address conflict detection

### LINKED ISSUE #80429: Bug in ACD implementation RFC5227
The ACD is not properly implemented as described in RFC5227 ch. 2.1.1
The implementation incorrectly detects an IP conflict, if an ARP request is received for the target IP.
The reason is that the current implementation checks for ARP requests instead of ARP probes.

To reproduce enable NET_IPV4_ACD and use DHCP to get an address. Immediately after address is bound send an arp request for the received IP. The DHCP server may already do that. Then a conflict is detected and a new IP requested and so on...

I fixed it with this change, which correctly checks for ARP probe instead of ARP request for the second condition:

```
--- zephyr/subsys/net/ip/ipv4_acd.c	2024-10-25 11:19:34.000000000 
+++ zephyr/subsys/net/ip/ipv4_acd.c	2024-10-24 16:00:44.000000000 
@@ -285,21 +285,25 @@
 			continue;
 		}
 
 		ll_addr = net_if_get_link_addr(addr_iface);
 
 		/* RFC 5227, ch. 2.1.1 Probe Details:
-		 * - Sender IP address match OR,
-		 * - Target IP address match with different sender HW address,
+		 * - ARP Request/Reply with Sender IP address match OR,
+		 * - ARP Probe where Target IP address match with different sender HW address,
 		 * indicate a conflict.
+		 * ARP Probe has an all-zero sender IP address
 		 */
+		uint32_t all_zero_ip = 0;
 		if (net_ipv4_addr_cmp_raw(arp_hdr->src_ipaddr,
 					  (uint8_t *)&ifaddr->address.in_addr) ||
 		    (net_ipv4_addr_cmp_raw(arp_hdr->dst_ipaddr,
 					  (uint8_t *)&ifaddr->address.in_addr) &&
-		     memcmp(&arp_hdr->src_hwaddr, ll_addr->addr, ll_addr->len) != 0)) {
+		     (memcmp(&arp_hdr->src_hwaddr, ll_addr->addr, ll_addr->len) != 0) &&
+		     (net_ipv4_addr_cmp_raw(arp_hdr->src_ipaddr,
+		                           (uint8_t *)&all_zero_ip)))) {
 			NET_DBG("Conflict detected from %s for %s",
 				net_sprint_ll_addr((uint8_t *)&arp_hdr->src_hwaddr,
 						   arp_hdr->hwlen),
 				net_sprint_ipv4_addr(&ifaddr->address.in_addr));
 
 			iface->config.ip.ipv4->conflict_cnt++;

```

### PR DESCRIPTION
**Pull request by anhuba**
The second condition needs to check ARP probes only

Fixes #80429

### CODE REVIEW

File: subsys/net/ip/ipv4_acd.c [hunk 1]
```diff
@@ -288,15 +288,19 @@ enum net_verdict net_ipv4_acd_input(struct net_if *iface, struct net_pkt *pkt)
...
 		 */
+		uint32_t all_zero_ip = 0;
 		if (net_ipv4_addr_cmp_raw(arp_hdr->src_ipaddr,
 					  (uint8_t *)&ifaddr->address.in_addr) ||
 		    (net_ipv4_addr_cmp_raw(arp_hdr->dst_ipaddr,
 					  (uint8_t *)&ifaddr->address.in_addr) &&
-		     memcmp(&arp_hdr->src_hwaddr, ll_addr->addr, ll_addr->len) != 0)) {
+		     (memcmp(&arp_hdr->src_hwaddr, ll_addr->addr, ll_addr->len) != 0) &&
+		     (net_ipv4_addr_cmp_raw(arp_hdr->src_ipaddr,
+		                           (uint8_t *)&all_zero_ip)))) {
```
Comment (jukkar) [inline subsys/net/ip/ipv4_acd.c:303]: We could avoid extra variable by this
```
(net_ipv4_addr_cmp_raw(arp_hdr->src_ipaddr,
	               (uint8_t *)(&(struct in_addr)INADDR_ANY_INIT)))) {
```

**Comment by jukkar** [comment]
Could you also add text from your issue to the commit message, at least this could be put there
```
The ACD is not properly implemented as described in RFC5227 ch. 2.1.1
The implementation incorrectly detects an IP conflict, if an ARP request is received for the target IP.
The reason is that the current implementation checks for ARP requests instead of ARP probes.
```

**Review by rlubos — COMMENTED** [review]
The fix itself looks good, but the branch needs some cleanup:
* Please use rebase instead of merge, no merge commits allowed
* The commit message in the fix commit is missing signoff entry.

**Comment by anhuba** [comment]
changed commit message and elminated local variable

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**